cpp/include/cudf/ast/detail/transform.cuh

```diff
@@ -317,16 +317,60 @@ __device__ void evaluate_row_expression(detail::row_evaluator const& evaluator,
   }
 }
 
+/**
+ * @brief The AST plan creates a device buffer of data needed to execute an AST.
+ *
+ * On construction, an AST plan creates a single "packed" host buffer of all necessary data arrays,
+ * and copies that to the device with a single host-device memory copy. Because the plan tends to be
+ * small, this is the most efficient approach for low latency.
+ *
+ * TODO: Remove comment below depending on final design
+ * The stream is not synchronized automatically, so a stream sync must be performed manually (or by
+ * another function) before the device data can be used safely.
+ *
+ */
 struct ast_plan {
  public:
-  ast_plan(linearizer const& expr_linearizer) : _sizes{}, _data_pointers{}
+  ast_plan(linearizer const& expr_linearizer,
+           rmm::cuda_stream_view stream,
+           rmm::mr::device_memory_resource* mr)
+    : _sizes{}, _data_pointers{}
   {
     add_to_plan(expr_linearizer.data_references());
     add_to_plan(expr_linearizer.literals());
     add_to_plan(expr_linearizer.operators());
     add_to_plan(expr_linearizer.operator_source_indices());
+
+    // Create device buffer
+    auto const h_data_buffer  = host_data_buffer();
+    auto const buffer_offsets = offsets();
+    auto const buffer_size    = h_data_buffer.second;
+    auto device_data_buffer =
+      rmm::device_buffer(h_data_buffer.first.get(), buffer_size, stream, mr);
+
+    // To reduce overhead, we don't call a stream sync here.
+    // The stream is synced later when the table_device_view is created.
+    // ^^^^ this comment will be removed, we are synchronizing vvvv
+    stream.synchronize();  // this doesn't seem to work
+
+    // Create device pointers to components of plan
+    auto const device_data_buffer_ptr = static_cast<const char*>(device_data_buffer.data());
+    _device_data_references           = reinterpret_cast<const detail::device_data_reference*>(
+      device_data_buffer_ptr + buffer_offsets[0]);
+    _device_literals = reinterpret_cast<const cudf::detail::fixed_width_scalar_device_view_base*>(
+      device_data_buffer_ptr + buffer_offsets[1]);
+    _device_operators =
+      reinterpret_cast<const ast_operator*>(device_data_buffer_ptr + buffer_offsets[2]);
+    _device_operator_source_indices =
+      reinterpret_cast<const cudf::size_type*>(device_data_buffer_ptr + buffer_offsets[3]);
   }
 
+  auto device_data_references() const { return _device_data_references; }
```

Comment:
I think it'd make sense to make these spans instead of raw pointers so you know the extent of them?

Reply:
I've chosen to just expose the `ast_plan` internals directly. I realize this probably runs counter to the typical preference for better encapsulation, but in this case I think it makes more sense since aside from the logic in its constructor we're basically treating the `ast_plan` as a trivial container for the various data pointers and getting all of them later to pass to the kernel. They're all const pointers, so there's little risk, and this is an internal-facing class so I don't think we need to be overly defensive.

I think we can hold off on making the internal variables spans. We could add later if needed (all the data is computed in the ctor) but for now there really isn't a need.

Reply:
How do you otherwise know the extent of these arrays? What's the concern with making them spans?

Reply:
No concern with spans, just a question of whether it's worth including without an immediate need since we'd also have to modify all the other device side code to accept spans and we aren't currently using the size. I can go ahead and do it if you prefer to future-proof and avoid passing around raw pointers.